src/libraries/System.Private.CoreLib/src/System/IO/BufferedFileStreamStrategy.cs

```diff
@@ -0,0 +1,132 @@
+
+        public override int ReadByte() => _bufferedStream.ReadByte();
+
+        public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state)
+            => _bufferedStream.BeginRead(buffer, offset, count, callback, state);
+
+        public override int EndRead(IAsyncResult asyncResult)
+            => _bufferedStream.EndRead(asyncResult);
+
+        public override int Read(byte[] buffer, int offset, int count) => _bufferedStream.Read(buffer, offset, count);
+
+        public override int Read(Span<byte> buffer) => _bufferedStream.Read(buffer);
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            => _bufferedStream.ReadAsync(buffer, offset, count, cancellationToken);
+
+        public override ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default)
+            => _bufferedStream.ReadAsync(buffer, cancellationToken);
+
+        public override IAsyncResult BeginWrite(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state)
+            => _bufferedStream.BeginWrite(buffer, offset, count, callback, state);
+
+        public override void EndWrite(IAsyncResult asyncResult)
+            => _bufferedStream.EndWrite(asyncResult);
+
+        public override void WriteByte(byte value) => _bufferedStream.WriteByte(value);
+
+        public override void Write(byte[] buffer, int offset, int count) => _bufferedStream.Write(buffer, offset, count);
+
+        public override void Write(ReadOnlySpan<byte> buffer) => _bufferedStream.Write(buffer);
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            => _bufferedStream.WriteAsync(buffer, offset, count, cancellationToken);
+
+        public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default)
+            => _bufferedStream.WriteAsync(buffer, cancellationToken);
+
+        public override void Flush() => _bufferedStream.Flush();
+
+        internal override void Flush(bool flushToDisk) => _bufferedStream.Flush(flushToDisk);
+
+        public override Task FlushAsync(CancellationToken cancellationToken)
+            =>  _bufferedStream.FlushAsync(cancellationToken);
+
+        public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+            => _bufferedStream.CopyToAsync(destination, bufferSize, cancellationToken);
+
+        public override ValueTask DisposeAsync()
+            => _bufferedStream.DisposeAsync();
+
+        internal override void DisposeInternal(bool disposing)
+        {
+            try
+            {
+                // the finalizer must at least try to flush the write buffer
+                // so we enforce it by passing always true
+                _bufferedStream.DisposeInternal(true);
```

Comment:
This seems... wrong.  I see that we "try" to flush today from the finalizer, but we also don't block waiting for that flush to complete, and the thing we're wrapping could have already been disposed of.

Reply:
Do you suggest that I should keep the old behavior of just starting the async flush in finalizer?

https://github.com/dotnet/runtime/blob/832a65623d180412fcbaefd68aa57065341db931/src/libraries/System.Private.CoreLib/src/System/IO/FileStream.Windows.cs#L348-L351

Reply:
@jkotas, how would you feel about FileStream not flushing its buffer at all during finalization?

Reply:
Flushing buffers during finalization was historically compensating for buggy apps that did not close the stream properly. We broke this to large degree when we stopped doing finalization during shutdown in .NET Core. 

There is still the case of the apps depending on in-flight finalization to flush the buffers. I expect a bug like this to be very rare. I think it is ok to break it if we have a good reason to. If we choose to do it, we should follow the process and file a breaking change template for it. Are there any other breaking changes coming with this rewrite?

Reply:
That's my thinking as well.

> Are there any other breaking changes coming with this rewrite?

Yes.  The plan for now is to have the old implementation available via a compat switch until we've proven to ourselves that the risk is low enough in removing it; I don't know yet whether that's before or after .NET 6 ships, but I expect after.